cmaes/_cma.py

```diff
@@ -324,6 +346,48 @@ def tell(self, solutions: List[Tuple[np.ndarray, float]]) -> None:
             + self._cmu * rank_mu
         )
 
+    def should_stop(self) -> bool:
+        B, D = self._eigen_decomposition()
+        dC = np.diag(self._C)
+
+        # Stop if the range of function values of the recent generation is below tolfun.
+        if (
+            self.generation > self._funhist_term
+            and np.max(self._funhist_values) - np.min(self._funhist_values)
+            < self._tolfun
+        ):
+            return True
+
+        # Stop if the std of the normal distribution is smaller than tolx
+        # in all coordinates and pc is smaller than tolx in all components.
+        if np.all(self._sigma * dC < self._tolx) and np.all(
+            self._sigma * self._pc < self._tolx
+        ):
+            return True
+
+        # Stop if detecting divergent behavior.
+        if self._sigma * np.max(D) > self._tolxup:
+            return True
+
+        # No effect coordinates: stop if adding 0.2-standard deviations
+        # in any single coordinate does not change m.
+        if np.any(self._mean == self._mean + (0.2 * self._sigma * np.sqrt(dC))):
+            return True
+
+        # No effect axis: stop if adding 0.1-standard deviation vector in
+        # any principal axis direction of C does not change m. "pycma" check
+        # axis one by one at each generation.
+        i = self.generation % self.dim
+        if np.all(self._mean == self._mean + (0.1 * self._sigma * D[i] * B[:, i])):
+            return True
```

Comment:
See the following implementation of pycma:

https://github.com/CMA-ES/pycma/blob/af8bd709c7982c288e364c99ddb22ad3c2946f0b/cma/evolution_strategy.py#L3483-L3501

<img width="897" alt="ScreenShot 2020-07-23 16 52 00" src="https://user-images.githubusercontent.com/5564044/88263370-e8f2a200-cd04-11ea-8c3f-24c2f390e3ae.png">